### PR TITLE
chore!: use torch and lightning 2.0

### PR DESCRIPTION
**Pull request by ab93**
- Support [pytorch 2.0](https://pytorch.org/blog/pytorch-2.0-release/)
- Speed up unit tests run
- Improve validation set run in data module
- Pin poetry version in github workflows

### CODE REVIEW

File: numalogic/models/autoencoder/base.py [hunk 1]
```diff
@@ -68,8 +85,10 @@ def configure_optimizers(self) -> Dict[str, Any]:
 
     def training_step(self, batch: Tensor, batch_idx: int) -> Tensor:
         loss = self._get_reconstruction_loss(batch)
+        self._total_train_loss += loss.detach().item()
```
Comment (qhuai) [inline numalogic/models/autoencoder/base.py:88]: as you only need a float value, do you still need the detach call? can you just do `loss.item()`?

File: numalogic/models/autoencoder/base.py [hunk 2]
```diff
@@ -29,6 +29,23 @@ def __init__(self, loss_fn: str = "huber", optim_algo: str = "adam", lr: float =
...
+        return self._total_train_loss
+
+    @property
+    def total_val_loss(self):
+        return self._total_val_loss
+
+    def reset_train_loss(self):
+        self._total_train_loss = 0.0
+
+    def reset_val_loss(self):
```
Comment (s0nicboOm) [inline numalogic/models/autoencoder/base.py:46]: Any specific reason we are introducing these?

File: numalogic/tools/callbacks.py [hunk 3]
```diff
@@ -40,12 +39,14 @@ def disable(self):
 
     def on_train_epoch_end(self, trainer: pl.Trainer, pl_module: pl.LightningModule) -> None:
         super().on_train_epoch_end(trainer, pl_module)
-        metrics = self.get_metrics(trainer, pl_module)
+        loss = pl_module.total_train_loss / trainer.num_training_batches
```
Comment (qhuai) [inline numalogic/tools/callbacks.py:42]: Is it guaranteed that `total_train_loss` is the result of `num_training_batches` additions?

File: pyproject.toml
Comment (s0nicboOm) [inline pyproject.toml:1]: We might need a minor update.

File: numalogic/tools/callbacks.py [hunk 4]
```diff
@@ -40,12 +39,14 @@ def disable(self):
...
         if curr_epoch % self._log_freq == 0:
-            _LOGGER.info("epoch=%s, training_loss=%s", curr_epoch, metrics["loss"])
+            _LOGGER.info("epoch=%s, training_loss=%.5f", curr_epoch, loss)
+        pl_module.reset_train_loss()
 
     def on_validation_epoch_end(self, trainer: pl.Trainer, pl_module: pl.LightningModule) -> None:
         super().on_validation_epoch_end(trainer, pl_module)
-        metrics = self.get_metrics(trainer, pl_module)
-        _LOGGER.info("validation_loss=%s", metrics["loss"])
+        loss = pl_module.total_val_loss / trainer.num_val_batches[0]
```
Comment (qhuai) [inline numalogic/tools/callbacks.py:50]: similar question applies here: about the consistency between total_val_loss and num_val_batches[0]

**Review by s0nicboOm — APPROVED** [review]

File: pyproject.toml
Comment (ab93) [reply pyproject.toml:1]: We are going to follow how numaflow keeps track of releases. `main` branch will have all the changes, and the necessary commits will be cherry-picked to `release/<version>` branches for package releases.
cc @whynowy @vigith

File: numalogic/tools/callbacks.py [hunk 5]
```diff
@@ -40,12 +39,14 @@ def disable(self):
...
+        _LOGGER.info("validation_loss=%.5f", loss)
```
Comment (qhuai) [inline numalogic/tools/callbacks.py:51]: Other than logging the validation loss, do we use the loss to drive other actions programmatically?

File: numalogic/models/autoencoder/base.py (hunk 2 above)
Comment (ab93) [reply numalogic/models/autoencoder/base.py:46]: Because before torch lightning 2.0 handled it automatically in the trainer. Now post 2.0, they are removing that, and giving more control to user so they have to define them. This was done to mostly speed up training and remove unnecessary overhead.

File: numalogic/models/autoencoder/base.py (hunk 1 above)
Comment (ab93) [reply numalogic/models/autoencoder/base.py:88]: In training, tensors are by default attached to the computational graph (autograd). We need to detach them first, before getting the value.

File: numalogic/tools/callbacks.py (hunk 5 above)
Comment (ab93) [reply numalogic/tools/callbacks.py:51]: Not now, but soon. We need a way to capture these losses and plot an error curve (train + val loss vs epochs).

File: numalogic/tools/callbacks.py (hunk 3 above)
Comment (ab93) [reply numalogic/tools/callbacks.py:42]: I didn't get your question here. `total_train_loss` is something we have defined, to capture the total loss in each mini-batch backpropagation. So mean loss across all batches can be gathered by dividing by `num_training_batches`.
